src/auth.ts

```diff
@@ -425,6 +425,67 @@ function urlsafeBase64(base64string: string) {
   return base64string.replace(/\+/g, "-").replace(/=+$/, "").replace(/\//g, "_");
 }
 
+interface PrototyperRes {
+  uri: string;
+  sessionId: string;
+  authorize: (authorizationCode: string) => Promise<UserCredentials>;
+}
+
+export async function loginPrototyper(): Promise<PrototyperRes> {
+  const authProxyClient = new apiv2.Client({
+    urlPrefix: authProxyOrigin(),
+    auth: false,
+  });
+
+  const sessionId = uuidv4();
+  const codeVerifier = randomBytes(32).toString("hex");
+  // urlsafe base64 is required for code_challenge in OAuth PKCE
+  const codeChallenge = urlsafeBase64(createHash("sha256").update(codeVerifier).digest("base64"));
+
+  const attestToken = (
+    await authProxyClient.post<{ session_id: string }, { token: string }>("/attest", {
+      session_id: sessionId,
+    })
+  ).body?.token;
+
+  const loginUrl = `${authProxyOrigin()}/login?code_challenge=${codeChallenge}&session=${sessionId}&attest=${attestToken}&studio_prototyper=true}`;
+  return {
+    uri: loginUrl,
+    sessionId: sessionId.substring(0, 5).toUpperCase(),
+    authorize: async (code: string) => {
+      try {
+        const tokens = await getTokensFromAuthorizationCode(
+          code,
+          `${authProxyOrigin()}/complete`,
+          codeVerifier,
+        );
+
+        const creds = {
+          user: jwt.decode(tokens.id_token!, { json: true }) as any as User,
```

Comment:
You should be able to use the generic type:
```suggestion
          user: jwt.decode<User>(tokens.id_token!, { json: true }),
```

Reply:
JWT doesn't have a generic typed version of decode

Reply:
Man, that's what I get for not researching an LLM response :( 
Can we re-type as `as unknown as User`? `unknown` is usually preferred over `any`